### PR TITLE
Range Field Histogram Aggregation

### PR DESCRIPTION
**Pull request by not-napoleon**
Work in progress

TODO:
[x] Range ValuesSource
[x] Fork the `HistogramAggregationFactory` to return different concrete classes based on field type (similar to Terms agg)
[x] Binary decode logic for ranges - see #41206 
[x] Implement RangeHistogramAggregation builder logic
[x] Implement RangeHistogramAggregation leaf collector logic
[x] Deal with IP range related edge cases
[ ] Yaml & Doc tests

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java]: Note to our future-selves: we should make sure this doesn't "bleed" over to other types that also use bytes... I think Strings will show up as `ValuesSource.Bytes` (from `ValueType.String` -> `ValuesSourceType.BYTES` -> `ValuesSource.Bytes`)

Unless that's handled elsewhere and we can ignore :)

Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java]: Oh, actually, we should have a more specific `ValuesSource` here, now that I've added a range specific choice.  Nice catch.

File: server/src/main/java/org/elasticsearch/index/fielddata/plain/DocValuesIndexFieldData.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/index/fielddata/plain/DocValuesIndexFieldData.java]: Note to self: @polyfractal and I kicked this around, and we think the correct short term solution (i.e. for this PR) is to add a `rangeType()` method or similar to flag this, similar to how numeric and script work, rather than relying on the field type.

File: server/src/main/java/org/elasticsearch/index/fielddata/plain/DocValuesIndexFieldData.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/index/fielddata/plain/DocValuesIndexFieldData.java]: This is currently only being used as a flag, but the overhead between storing a boolean and storing an enum reference isn't high, so it seemed worth leaving ourselves access to the more robust data.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java]: The type hierarchy in ValuesSource is a little arcane, and I just picked something that looked reasonable.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java]: I'm almost positive this is the wrong thing to do here, but I'm not sure what the right thing to do is.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: Perhaps pull the from/to out of the loop into temp variables, just to make the line a little shorter/easier to read?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: We could probably add a `doCollect()` method on some kind of super class that both extend... but if this is the only place where they share code it's probably not worth the extra complexity.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Maybe promote this up to an `else if`?  Not sure if it'd be cleaner or not, but it's nice to reduce nested if where possible.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: So this is actually the case where the field is unmapped _and_ the user has specified a [`missing` parameter](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-histogram-aggregation.html#_missing_value_13) to fill in missing values with some kind of default.

Assuming we support the `missing` param on ranges too, I think we'll need to set a valid VS here.  We'll also need something down below that does the equivalent of `MissingValues.replaceMissing()` with the user-supplied range.

That said... I'm not entirely sure how or if this works with ranges.  The `missing` param in other aggs only deals with a string or numeric, not a complex to/from object.  So perhaps we just don't support `missing`, but in that case we'll have to make sure it's [not added to the parser](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceParserHelper.java#L68-L69) for ranges.  I'm not opposed to skipping support for `missing` on first iteration and adding it later if requested by users.  Being a runtime feature it won't have any BWC implications and we can add later if desired.


(The unmapped (but no `missing`) scenario is handled above where vs is set to `null`, which should be fine.)

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Probably, eventually.  I suspect it's non-trivial work though, so OK marking that as a limitation for first pass and we can revisit.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java]: Just a thought: do you know if we have a test anywhere that validates these enum orderings don't change?  We have a test class for that (`AbstractWriteableEnumTestCase`), but I think there are some custom one-off implementations too.

If we don't have a test, might be a nice time to add one just to ensure these don't get broken in the future.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java
Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java]: Yeah, so this is a bit of a split personality in the framework right now.  Setting the valuesSource to null will use a no-op collector in the histo's `getLeafCollector()`.  Then it uses the same histo aggregator to build an empty result.

Terms agg (and a few others) instead have an "unmapped" version of the agg (`UnmappedTerms`) which is used instead.  This unmapped version uses the same no-op collector, and knows how to build an empty response.

I'm not sure if there is a preference or why it's a split-personality.  I've pinged @colings86 to see if he knows the history here :)

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java
Comment (polyfractal) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java]: We should probably beef these tests up.  E.g. some more "good" scenarios, testing out generic params like min_doc_count/offset/ordering, different datatypes, expected exceptions (illegal intervals, etc)

If you're still working on tests, ignore this. :)

**Review by polyfractal — COMMENTED** [review]
Looking good!  Did a pass and left some comments/questions, and a few style nits (sorry!). :)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: They share this and the `buildAggregtion` & `buildEmptyAggregation` methods below.  I didn't feel it was worth the complexity when I wrote it, which is why it's copied here instead of in a super class, but I was a little on the fence about it.  I was thinking to see what the overall framework refactoring plan looked like, and if there might be a logical place or something like this there.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Sure.  The reason I had an extra level of nesting here was to emphasize that we were checking `fieldType` instead of `indexFieldData` like the other clauses do, but reading it now I'm not sure that's necessary to call out at all, and even if it is, a comment would do better than some fancy-pants nesting.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java
Comment (not-napoleon) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregatorTests.java]: Yeah, I'm still working on tests & docs :)

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java]: Having just done some more thinking on the `ValuesSource` refactor, I'm in agreement that Range should be a peer to `Numeric`, `GeoPoint`, etc.  I'll wire that up shortly.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java]: I added a note about this in #42949 (aka the ValuesSource refactor ticket).  I don't think this PR is the best place to add a serialization test for this enum though.  If you think it's high urgency, I'm happy to open a different PR against master to address it, or deal with it as part of the refactor.

Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java]: :+1: Let's do it as a separate PR.  I don't think it's super urgent, but untested enums make me nervous so doing it as a separate PR before tackling the refactor is probably a good idea.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: We do not currently check this before getting to this point.  I don't see a reason we couldn't move it up, I just put it here to mirror how the agg worked before I touched it.

Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: :+1:

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: ^^^ ?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: Will things go pear shaped if we ever get non-sorted ranges?  Should we toss an assertion in here?

Or does it not particularly matter what order the ranges come in?

**Review by polyfractal — COMMENTED** [review]
Left two questions, I think this is good to go once the ANY/missing stuff is integrated!  Gimme a ping once that gets merged into this PR and I'll do a quick skim. :+1:

**Comment by not-napoleon** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by not-napoleon** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Don't think we need this comment anymore?

File: server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java]: Ditto, think this is outdated?

**Review by polyfractal — APPROVED** [review]
Think there are a few out-of-date comments, and curious about comments left in last review for `RangeHistogramAggregator` (about assertions, sorted input), but otherwise LGTM! :tada:

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/RangeHistogramAggregator.java]: It matters, but the encoding sorts them.  Doesn't hurt to throw an assert in though, I'll do that.

**Comment by polyfractal** [comment]
:tada:

**Pull request merged**
